content/hacking-atom/sections/creating-a-grammar.md

```diff
@@ -0,0 +1,186 @@
+---
+title: Creating a Grammar
+---
+### Creating a Grammar
+
+Grammars power Atom's syntax highlighting. Each grammar provides a set of _regex_ patterns which are assigned _scopes_. These scopes are then turned into the CSS classes that you can target in syntax themes!
```

Comment:
For the absolute newbies, maybe this should be "*regex* (regular expression) patterns". The only mention of the full name is two paragraphs down in a URL.

Reply:
If you hope to provide for newbies, you will have to have text that starts at first in terms even more basic than scopes and css classes. Maybe best to show first what it can do and how you can do it, without the technicalities. Get the 'client' interested first and then follow up by terminology IMHO.

Reply:
I don't think it needs to target people who don't know what CSS classes are. Obviously, it *could*, but how many people at that level of familiarity are going to be making a language package?

Reply:
> how many people at that level of familiarity are going to be making a language package?
At least one... **_me_**. As I am not so special, there should be more people like that.

Following your logic, the people creating a grammar should then be people who knows regular expressions or regex or RE. There should not even be a web link to what it is and what it does.

\- thinking aloud - the double escaping in the regular expression patterns was very much a surprise.

Reply:
There are plenty of tutorial resources for CSS that handle it much better than an off-hand explanation in a grammar tutorial would, and I would have anyone who doesn't know CSS read a couple of them before starting. If you need, I can direct them to that. Note that I also didn't recommend that we instruct people on how to use regular expressions, only that we should explain what "regex" is short for. This is because there are plenty of regex tutorials. We could link to both kinds of tutorials, but we shouldn't devote paragraphs to them.

> - thinking aloud - the double escaping in the regular expression patterns was very much a surprise.

This is a unique quirk, based on the fact that CSON has to be parsed into JSON, and isn't applicable under the circumstance where the package author chooses to write in JSON. It definitely deserves a line or two, because it's not expected behavior, but it would never be part of any standard regex tutorial.